src/shared/usbspec.rs

```diff
@@ -0,0 +1,967 @@
+pub const USB_ENDPOINT_TYPE_ISOCHRONOUS_SYNCHRONIZATION_NO_SYNCHRONIZATION: UCHAR = 0x00;
+pub const USB_ENDPOINT_TYPE_ISOCHRONOUS_SYNCHRONIZATION_ASYNCHRONOUS: UCHAR = 0x04;
+pub const USB_ENDPOINT_TYPE_ISOCHRONOUS_SYNCHRONIZATION_ADAPTIVE: UCHAR = 0x08;
+pub const USB_ENDPOINT_TYPE_ISOCHRONOUS_SYNCHRONIZATION_SYNCHRONOUS: UCHAR = 0x0C;
+#[inline]
+pub fn USB_ENDPOINT_TYPE_ISOCHRONOUS_SYNCHRONIZATION(bmAttr: UCHAR) -> UCHAR {
+    bmAttr & USB_ENDPOINT_TYPE_ISOCHRONOUS_SYNCHRONIZATION_MASK
+}
+pub const USB_ENDPOINT_TYPE_ISOCHRONOUS_USAGE_MASK: UCHAR = 0x30;
+pub const USB_ENDPOINT_TYPE_ISOCHRONOUS_USAGE_DATA_ENDOINT: UCHAR = 0x00;
+pub const USB_ENDPOINT_TYPE_ISOCHRONOUS_USAGE_FEEDBACK_ENDPOINT: UCHAR = 0x10;
+pub const USB_ENDPOINT_TYPE_ISOCHRONOUS_USAGE_IMPLICIT_FEEDBACK_DATA_ENDPOINT: UCHAR = 0x20;
+pub const USB_ENDPOINT_TYPE_ISOCHRONOUS_USAGE_RESERVED: UCHAR = 0x30;
+#[inline]
+pub fn USB_ENDPOINT_TYPE_ISOCHRONOUS_USAGE(bmAttr: UCHAR) -> UCHAR {
+    bmAttr & USB_ENDPOINT_TYPE_ISOCHRONOUS_USAGE_MASK
+}
+STRUCT!{#[repr(packed)] struct USB_HIGH_SPEED_MAXPACKET_MP {
+    BitField: USHORT,
+}}
+BITFIELD!{USB_HIGH_SPEED_MAXPACKET_MP BitField: USHORT [
+    MaxPacket set_MaxPacket[0..11],
+    HSmux set_HSmux[11..13],
+    Reserved set_Reserved[13..16],
+]}
+UNION2!{#[repr(packed)] union USB_HIGH_SPEED_MAXPACKET {
+    [u16; 1],
+    s s_mut: USB_HIGH_SPEED_MAXPACKET_MP,
+    us us_mut: USHORT,
+}}
+pub type PUSB_HIGH_SPEED_MAXPACKET = *mut USB_HIGH_SPEED_MAXPACKET;
+pub const USB_ENDPOINT_SUPERSPEED_BULK_MAX_PACKET_SIZE: USHORT = 1024;
+pub const USB_ENDPOINT_SUPERSPEED_CONTROL_MAX_PACKET_SIZE: USHORT = 512;
+pub const USB_ENDPOINT_SUPERSPEED_ISO_MAX_PACKET_SIZE: USHORT = 1024;
+pub const USB_ENDPOINT_SUPERSPEED_INTERRUPT_MAX_PACKET_SIZE: USHORT = 1024;
+STRUCT!{#[repr(packed)] struct USB_STRING_DESCRIPTOR {
+    bLength: UCHAR,
+    bDescriptorType: UCHAR,
+    bString: [WCHAR; 1],
+}}
+pub type PUSB_STRING_DESCRIPTOR = *mut USB_STRING_DESCRIPTOR;
+pub const MAXIMUM_USB_STRING_LENGTH: UCHAR = 255;
+STRUCT!{#[repr(packed)] struct USB_SUPERSPEED_ENDPOINT_COMPANION_DESCRIPTOR_bmAttributes_Bulk {
+    BitField: UCHAR,
+}}
+BITFIELD!{USB_SUPERSPEED_ENDPOINT_COMPANION_DESCRIPTOR_bmAttributes_Bulk BitField: UCHAR [
+    MaxStreams set_MaxStreams[0..5],
+    Reserved1 set_Reserved1[5..8],
+]}
+STRUCT!{#[repr(packed)]
+    struct USB_SUPERSPEED_ENDPOINT_COMPANION_DESCRIPTOR_bmAttributes_Isochronous {
+    BitField: UCHAR,
+}}
+BITFIELD!{USB_SUPERSPEED_ENDPOINT_COMPANION_DESCRIPTOR_bmAttributes_Isochronous BitField: UCHAR [
+    Mult set_Mult[0..2],
+    Reserved2 set_Reserved2[2..7],
+    SspCompanion set_SspCompanion[7..8],
+]}
+UNION2!{#[repr(packed)] union USB_SUPERSPEED_ENDPOINT_COMPANION_DESCRIPTOR_bmAttributes {
+    [u8; 1],
+    AsUchar AsUchar_mut: UCHAR,
+    Bulk Bulk_mut: USB_SUPERSPEED_ENDPOINT_COMPANION_DESCRIPTOR_bmAttributes_Bulk,
+    Isochronous Isochronous_mut:
```

Comment:
In this case, the bitfields and unions are fine as is because the bitfield structs' fields are named.

Reply:
How about removing intermediate struct for bitfield and applying bitfields macro directly to Bulk or Isochronous variable?

Reply:
Because of how the bitfields are accessed in C. `.bmAttributes.Bulk.MaxStreams` for example. If the intermediate struct is used in a *named* field, then in C you must go through that name to access the bitfields, so it must be retained in Rust for a similar experience.